src/ctypes/ldouble_stubs.c

```diff
@@ -0,0 +1,525 @@
+/*
+ * Copyright (c) 2016 Andy Ray.
+ *
+ * This file is distributed under the terms of the MIT License.
+ * See the file LICENSE for details.
+ */
+
+#if !__USE_MINGW_ANSI_STDIO && (defined(__MINGW32__) || defined(__MINGW64__))
+#define __USE_MINGW_ANSI_STDIO 1
+#endif
+
+#include <caml/mlvalues.h>
+#include <caml/custom.h>
+#include <caml/alloc.h>
+#include <caml/intext.h>
+#include <caml/fail.h>
+#include <caml/hash.h>
+#include <caml/memory.h>
+
+#include <stdio.h>
+#include <stdint.h>
+#include <float.h>
+#include <math.h>
+#include <complex.h>
+
+#include "ctypes_ldouble_stubs.h"
+
+/*********************** long double *************************/
+
+/*
+ * long double comes in various different flavours on different
+ * platforms/architectures.
+ *
+ * 8 byte double - arm, msvc
+ * 10 byte extended - intel gcc.  can be packed into 12 or 16 bytes.
+ * 16 byte - powerpc, either IEEE quad float or __ibm128 double double
+ *
+ * We make a best guess as to the format based on LDBL_MANT_DIG. 
+ * This only affects the operation of hashing and serialization.
+ *
+ * For deserialization we consider it an error if the stored
+ * value is a different format.  Doing such conversions would
+ * get very complicated.
+ *
+ * Regarding endianness - the 8 and 16 byte formats should
+ * interwork between big and little endian systems.  The
+ * intel extended 10 byte format only seems to occurs on 
+ * x86 so we dont need to consider endianness.
+ *
+ * In case a format is encountered that we do not understand, 
+ * then we fall back to casting the value to a double.
+ *
+ */
+
+#define LDOUBLE_STORAGE_BYTES sizeof(long double)
+#if (LDBL_MANT_DIG == 53)      // 64 bit - same as double
+#define LDOUBLE_VALUE_BYTES 8
+#elif (LDBL_MANT_DIG == 64)    // intel 80 bit extended 
+#define LDOUBLE_VALUE_BYTES 10
+#elif (LDBL_MANT_DIG == 106)   // __ibm128 (pair of doubles)
+#define LDOUBLE_VALUE_BYTES 16
+#elif (LDBL_MANT_DIG == 113)   // ieee __float128
+#define LDOUBLE_VALUE_BYTES 16
+#else
+#define LDOUBLE_VALUE_BYTES LDOUBLE_STORAGE_BYTES
+#endif
+
+#define ldouble_custom_val(V) (*(long double *)(Data_custom_val(V)))
+
+static int ldouble_cmp(long double u1, long double u2) {
+  return (u1 == u2) ? 0 : (u1 < u2 ? -1 : 1);
```

Comment:
I see a couple of ways in which the current `Ldouble` implementation diverges from the behaviour for OCaml `float`:

* With `float`, all NaNs are equal under `compare`:
  ```ocaml
  # compare (float_of_string "nan") (float_of_string "nan") ;;
  - : int = 0
  ```
  but with `ldouble` distinct NaNs are unequal under `compare`:
  ```ocaml
  # compare (Ldouble.of_string "nan") (Ldouble.of_string "nan") ;;
  - : int = 1
  ```

* With `float` NaNs are less than all other numbers under `compare`
  ```ocaml
  # [compare nan 1.0; compare nan (-1.0); compare nan neg_infinity];;
  - : int list = [-1; -1; -1]
  ```
  but with `ldouble` NaNs are greater than other numbers under `compare`: 
  ```ocaml
  # Ldouble.[compare nan (of_float 1.0);
             compare nan (of_float (-1.0));
             compare nan neg_infinity];;
  - : int list = [1; 1; 1]
  ```
  
Replicating the OCaml behaviour may require setting `caml_compare_unordered`, as in the [bigarray stubs](https://github.com/ocaml/ocaml/blob/f68ff2e5b0bd1ba3efe37ec336c02226237cc48b/otherlibs/bigarray/bigarray_stubs.c#L734-L747).